### ISSUE TITLE
Модуль 1. 

### ISSUE THREAD

**Issue by zhukovsky** [CV-NLP-Lab/SciDev#1]
Реализовать (и проверить!) алгоритмы, требуемые для обучения нейронной сети.
### 1. Алгоритмы оптимизации
Файл: optimizers.py
Требуется реализовать следующие алгоритмы обучения нейронной сети:
1. Градиентный спуск (class GradientDescent) (Лекция 2) с моментом (включая Нестерова) и регуляризацией весов (L1, L2) (Лекция 3)
2. Adam (Лекция 3)
_Реализация алгоритма должна не зависеть от размера батча при стохастическом поведении_

### 2. Обучение многослойной модели, обратное распространение ошибки
models.py: class FeedForwardModel

```
__init__ -- конструктор
forward -- прямой проход сети
backward -- обратный проход сети, один 
zero_grad -- обнуление всех накопленных градиентов
apply_grad -- модификация весов, исходя из посчитанных градиентов
train -- цикл обучения. Параметры:
- data -- данные для обучения
- n_epochs -- количество эпох обучения
- batch_size -- размер батча для обучения
```
Алгоритм обучения должен быть независим от слоев, входящих в модель (Лекция 2). Код, зависимый от конкретного слоя или алгоритма оптимизации должен быть реализован в классе, соответствующем данному слою или алгоритму оптимизации.

Алгоритм обучения модели должен поддерживать стохастическое поведение с заданным размером батча (Лекция 4)

### 3. Архитектура сети
Требуется реализовать следующие элементы сети:
1. Линейный слой. Должен поддерживать вычисление от тензора произвольного размера, без требования приведения его к вектору (Лекция 2)
2. Функции активации: relu, сигмоида (Лекция 2)
3. Функции потерь: MSE (Лекция 1), softmax (Лекция 4), кросс-энтропия и KL-дивергенция (Лекция 4)

### 4. Проверка работоспособности сети
1. Алгоритм линейной регрессии. Один линейный слой с одним весом, без сдвига. Функция потерь — MSE, обучение с помощью классического градиентного спуска. Значение веса и функции потерь должно точно совпадать с версией из pytorch.
2. Модель, реализующая классификацию на датасете iris. Используемые составляющие и параметры обучения/архитектуры ограничены только реализованными Вами методами. Качество работы оценивается по доле правильных ответов (sklearn.accuracy_score). Качество работы должно быть сопоставимо с sklearn.logistic_regression. Все проведенные эксперименты (в том числе и неудачные!) и сделанные из них выводы необходимо записать.
